### ISSUE TITLE
다양한 머신 러닝 모델로 훈련한 결과 (GM, LGBM, RFR)

### ISSUE THREAD

**Issue by UGeunJi** [UGeunJi/Seoul_Bike_Rental_Prediction#7]
<html>
<body>
<pre><code class="fenced-code-block language-python">import pandas as pd
import random
import os
import numpy as np
import warnings
warnings.filterwarnings(&#39;ignore&#39;)

from sklearn.linear_model import LinearRegression
from sklearn.multioutput import MultiOutputRegressor
from sklearn import metrics

from sklearn.datasets import make_regression
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import train_test_split

from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression

# lightgbm for regression
from numpy import mean
from numpy import std
from lightgbm import LGBMRegressor
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedKFold
from matplotlib import pyplot</code></pre>
<h3 id="예측률을-향상시키기-위해-데이터를-무작위로-섞어줌" class="atx">예측률을 향상시키기 위해 데이터를 무작위로 섞어줌</h3>
<pre><code class="fenced-code-block language-python">def seed_everything(seed):
    random.seed(seed)
    os.environ[&#39;PYTHONHASHSEED&#39;] = str(seed)
    np.random.seed(seed)
seed_everything(42) # Seed 고정</code></pre>
<h2 id="데이터-수집" class="atx">데이터 수집</h2>
<pre><code class="fenced-code-block language-python">train = pd.read_csv(&#39;data/train_null_only.csv&#39;)
test = pd.read_csv(&#39;data/test_null_only.csv&#39;)
submission = pd.read_csv(&#39;data/submission.csv&#39;)</code></pre>
<pre><code class="fenced-code-block language-python">train.head()</code></pre>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

<pre><code class="fenced-code-block">.dataframe tbody tr th {
    vertical-align: top;
}

.dataframe thead th {
    text-align: right;
}</code></pre>
<p></style></p>

  | id | hour | hour_bef_temperature | hour_bef_precipitation | hour_bef_windspeed | hour_bef_humidity | hour_bef_visibility | hour_bef_ozone | hour_bef_pm10 | hour_bef_pm2.5 | count
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
3 | 20 | 16.3 | 1.0 | 1.5 | 89.0 | 576.0 | 0.027 | 76.0 | 33.0 | 49.0
6 | 13 | 20.1 | 0.0 | 1.4 | 48.0 | 916.0 | 0.042 | 73.0 | 40.0 | 159.0
7 | 6 | 13.9 | 0.0 | 0.7 | 79.0 | 1382.0 | 0.033 | 32.0 | 19.0 | 26.0
8 | 23 | 8.1 | 0.0 | 2.7 | 54.0 | 946.0 | 0.040 | 75.0 | 64.0 | 57.0
9 | 18 | 29.5 | 0.0 | 4.8 | 7.0 | 2000.0 | 0.057 | 27.0 | 11.0 | 431.0

</div>

<pre><code class="fenced-code-block language-python">train.shape, test.shape, submission.shape</code></pre>
<pre><code class="fenced-code-block">((1459, 11), (715, 10), (715, 2))</code></pre>
<h2 id="데이터-준비" class="atx">데이터 준비</h2>
<pre><code class="fenced-code-block language-python">features = [&#39;hour&#39;, &#39;hour_bef_temperature&#39;, &#39;hour_bef_precipitation&#39;, &#39;hour_bef_windspeed&#39;, &#39;hour_bef_humidity&#39;, &#39;hour_bef_visibility&#39;, 
            &#39;hour_bef_ozone&#39;, &#39;hour_bef_pm10&#39;, &#39;hour_bef_pm2.5&#39;]
X_train = train[features]
y_train = train[&#39;count&#39;]
X_test = test[features]</code></pre>
<pre><code class="fenced-code-block language-python">print(X_train.shape)
print(y_train.shape)
print(X_train.shape)</code></pre>
<pre><code class="fenced-code-block">(1459, 9)
(1459,)
(1459, 9)</code></pre>
<h2 id="gradientboostingregressor" class="atx">GradientBoostingRegressor</h2>
<p><a href="https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html">https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html</a></p>
<pre><code class="fenced-code-block language-python">GB = (GradientBoostingRegressor()).fit(X_train, y_train)

GB_pred = GB.predict(X_test)</code></pre>
<pre><code class="fenced-code-block language-python">submission[&#39;count&#39;] = GB_pred
submission.to_csv(&quot;GB_pred.csv&quot;, index = False)</code></pre>
<h2 id="randomforestregressor" class="atx">RandomForestRegressor</h2>
<p><a href="https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html#sklearn.ensemble.RandomForestRegressor">https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html#sklearn.ensemble.RandomForestRegressor</a></p>
<pre><code class="fenced-code-block language-python">X, y = make_regression(n_features=4, n_informative=2, random_state=0, shuffle=False)
regr = RandomForestRegressor(max_depth=2, random_state=0)
RFR = regr.fit(X_train, y_train)

RFR_pred = RFR.predict(X_test)</code></pre>
<pre><code class="fenced-code-block language-python">submission[&#39;count&#39;] = RFR_pred
submission.to_csv(&quot;RFR_pred.csv&quot;, index = False)</code></pre>
<h2 id="lightgradientboostingregressor" class="atx">LightGradientBoostingRegressor</h2>
<p><a href="https://nicola-ml.tistory.com/51">https://nicola-ml.tistory.com/51</a></p>
<pre><code class="fenced-code-block language-python">LGBM = (LGBMRegressor()).fit(X_train, y_train)

LGBM_pred = LGBM.predict(X_test)</code></pre>
<pre><code class="fenced-code-block language-python">submission[&#39;count&#39;] = LGBM_pred
submission.to_csv(&quot;LGBM_pred.csv&quot;, index = False)</code></pre>

</body>
</html>

## 제출 결과

![image](https://user-images.githubusercontent.com/84713532/204138281-f322f14f-3f99-4132-80b9-0c7bbfd79698.png)

- 세 모델 중에선 LGBM이 가장 높은 것을 알 수 있었다.
- RFR는 파라미터를 조정해 봤는데 결과가 처참한 것을 볼 수 있다. 애매하게 건드리면 안된다는 것을 배웠다.
- 이전에 데이터셋인 파라미터를 거의 조정하지 않은 RFR의 에러가 가장 낮은 것을 볼 수 있었는데(score: 45.1xxxxxx), 이를 통해 다양한 머신 러닝 모델을 사용하거나 파라미터를 조절하는 것보다 데이터 전처리를 해봐야 겠다고 생각하게 되었다.
